# Sowing Success: How Machine Learning Helps Farmers Select the Best Crops

![Farmer in a field](farmer_in_a_field.jpg)

Measuring essential soil metrics such as nitrogen, phosphorous, potassium levels, and pH value is an important aspect of assessing soil condition. However, it can be an expensive and time-consuming process, which can cause farmers to prioritize which metrics to measure based on their budget constraints.

Farmers have various options when it comes to deciding which crop to plant each season. Their primary objective is to maximize the yield of their crops, taking into account different factors. One crucial factor that affects crop growth is the condition of the soil in the field, which can be assessed by measuring basic elements such as nitrogen and potassium levels. Each crop has an ideal soil condition that ensures optimal growth and maximum yield.

A farmer reached out to you as a machine learning expert for assistance in selecting the best crop for his field. They've provided you with a dataset called `soil_measures.csv`, which contains:

- `"N"`: Nitrogen content ratio in the soil
- `"P"`: Phosphorous content ratio in the soil
- `"K"`: Potassium content ratio in the soil
- `"pH"` value of the soil
- `"crop"`: categorical values that contain various crops (target variable).

Each row in this dataset represents various measures of the soil in a particular field. Based on these measurements, the crop specified in the `"crop"` column is the optimal choice for that field.  

In this project, you will build multi-class classification models to predict the type of `"crop"` and identify the single most importance feature for predictive performance.

In [9]:
# All required libraries are imported here for you.
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn import metrics

# Load the dataset
crops = pd.read_csv("soil_measures.csv")

crops.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2200 entries, 0 to 2199
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   N       2200 non-null   int64  
 1   P       2200 non-null   int64  
 2   K       2200 non-null   int64  
 3   ph      2200 non-null   float64
 4   crop    2200 non-null   object 
dtypes: float64(1), int64(3), object(1)
memory usage: 86.1+ KB


every column has 2200 non-null lines

In [10]:
crops['crop'].value_counts()

rice           100
maize          100
jute           100
cotton         100
coconut        100
papaya         100
orange         100
apple          100
muskmelon      100
watermelon     100
grapes         100
mango          100
banana         100
pomegranate    100
lentil         100
blackgram      100
mungbean       100
mothbeans      100
pigeonpeas     100
kidneybeans    100
chickpea       100
coffee         100
Name: crop, dtype: int64

the classes are balanced

In [11]:
# Separate features (all columns except 'crop') and target (the 'crop' column)
X = crops.drop(columns='crop')  # Features
y = crops['crop']  # Target

# Dictionary to store performance for each feature
feature_performance = {}

# Iterate over each feature
for feature in X.columns:
    # Select the feature
    X_feature = X[[feature]]  # Selecting only one feature
    
    # Split data into train and test sets
    X_train, X_test, y_train, y_test = train_test_split(X_feature, y, test_size=0.2, random_state=42)
    
    # Train logistic regression model
    model = LogisticRegression(random_state=42)
    model.fit(X_train, y_train)
    
    # Make predictions
    y_pred = model.predict(X_test)
    
    # Evaluate the performance using F1-score for multi-class classification
    f1_score = metrics.f1_score(y_test, y_pred, average='weighted')
    
    # Store the performance in the dictionary
    feature_performance[feature] = f1_score

# Display the dictionary with feature performance
print(feature_performance)

{'N': 0.09149868209906838, 'P': 0.14761942909728204, 'K': 0.23896974566001802, 'ph': 0.04532731061152114}


## Interpretation of Results:
- **Potassium (K)** has the highest F1-score (0.239), meaning it seems to be the most predictive feature for crop classification in your dataset.
- **Phosphorous (P)** comes next with an F1-score of 0.148, followed by **Nitrogen (N)** with 0.091.
- **pH** has the lowest F1-score (0.045), indicating that it contributes the least to crop prediction in this particular case.

In [12]:
# Dictionary to store performance for each feature
best_predictive_feature = {}

# Find the best predictive feature (the one with the highest F1-score)
best_feature = max(feature_performance, key=feature_performance.get)
best_predictive_feature[best_feature] = feature_performance[best_feature]

# Display the dictionary with the best predictive feature
print("Best Predictive Feature and Score:")
print(best_predictive_feature)

Best Predictive Feature and Score:
{'K': 0.23896974566001802}
